src/lib/useReactDataTableState/useReactDataTableState.ts

```diff
@@ -1,18 +1,20 @@
 ﻿import { useState } from "react";
-import { ColumnFiltersState, PaginationState, SortingState } from "@tanstack/react-table";
+import { PaginationState } from "@tanstack/react-table";
 import { useReactDataTableStateProps } from "./useReactDataTableStateProps";
 import { useReactDataTableStateResult } from "./useReactDataTableStateResult";
+import { FilterModel } from "../types/TableState";
+import { SortingState } from "../types/SortingState";
 
 /**
  * A custom hook that will initialize all the state needed for the react data table
  * @returns the state and the setters
  */
-const useReactDataTableState = (props: useReactDataTableStateProps = {}): useReactDataTableStateResult => {
+const useReactDataTableState = <TData, TFilter extends FilterModel>(props: useReactDataTableStateProps<TData, TFilter> = {}): useReactDataTableStateResult<TData, TFilter> => {
   const { initialColumnFilters, initialSorting, initialPagination } = props;
 
-  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>(initialColumnFilters ?? []);
-  const [afterSearchFilter, setAfterSearchFilter] = useState<ColumnFiltersState>(initialColumnFilters ?? []);
-  const [sorting, setSorting] = useState<SortingState>(initialSorting ?? []);
+  const [columnFilters, setColumnFilters] = useState<TFilter>(initialColumnFilters as TFilter);
+  const [afterSearchFilter, setAfterSearchFilter] = useState<TFilter>(initialColumnFilters as TFilter);
+  const [sorting, setSorting] = useState<SortingState<TData>>(initialSorting as SortingState<TData>);
```

Comment:
@neotob @neotrow Hi guys, This cast for type is obviously wrong, I used it for testing. 

What I noticed is that, if I assign the correct type `TFilter | undefined` and `SortingState<TData> | undefined` we will work in all projects with states that are always undefined. 

I don't have a solution for this, what I was thinking is to have separated hooks like useSortingState, useColumnFiltersState and usePaginationState and to change this one to have all props as required. So that the user can use the one he wants separately or this one including all stuffs with the correct type.
What do you think?

Reply:
if i understand it correct, then either the TFilter must be able to be initiated with `{}` (all props must be nullable) or an initial state is mandatory.. else the base idea of this hook is not working 🤔 

idk if there is a way (either compile or runtime) to secure this? because if this would be given, we could do something like `(initialColumnFilters || {}) as TFilter` which then would be correct, right?

Reply:
@neotob I thought also to your suggestion, but I'm not really convinced. 

This is because if you have an object like `{ test: string }` and you pass to the useReactDataTable undefined, you will have back an object `{} ` which is misleading in TS as you will expect the test prop as defined, but then doing like `myObj.test.splice()` it will crash for undefined reference. Do you think that this is not a great issue as the end you should have passed it before?

Reply:
I tried to experiment a little around.. maybe i was able to create something like a possible solution: [playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoQM5gMpjpAQQBsiA5AVxLgCMiUBvAWAChk3kAHKAew4EYA-AC5k1btzpwQAbhbtOPDgCZhyEOQC21aMlnMAvixahIsRCnRZc+CKW5hiZSkRp1kTVuy69BI9Vug9eW9lESwoUABzPUNmFjAATw4UAAVgBABrCipaCAAeVAA+ZABedzlkAG0U5FBkDIgE7hg0ZDgMNWdkCAAPSBAAEw7UaoBdZAFkGr8IADdoUZERlNGWWPik1PSs+2yXXILiso82atqQesbm1vbOkm6+iEHhsYm1OZ0RFaWxtaNmRLJZAAeQ4YGA3BAcCcOToh1K5U8lQA0udLk0WmlMntXPkiqNVABRHoIIjkAZ4lGjAA0dyIhTWyAAZIjTqi6g0MVNtnYwDiDvifsjVgZ-oCUI5+XCigiTuwUWjOddUItkORBhAYKAIAMHv0hmgqW8RsLkDN5lAYnoNkDrJB4WVUHqngbJc5cfDJgAKJ0AHzVGq1IB1yH96hIAEozWhrcwTNB4EhkABxCBgAAimrglBweEgKUUGHhcrYoGA4OhdogIirhxi-xg6oQ4MhyEiacz8BztaKXpCGBEoJbUJh+zhqYzWe7eYgBd4RaKhQjS1ZyAQkKw7nO5eAlZnyH0CP7bQ6E872aIuZsc44C8KQWQUDT5CgFzLFaIVYmk3oh9uqHrOJmHXEAMAkCAADoiG4SIvXbScu0vHtMCvQgSClCBCi9egDwjCM9BAsC6CgmC4I7KckJnAoUKrN1YUw7Dtw-KsRHVCkgxDfQ8IIjdwJI2D4PPacbGoqwZzoscGJw99d0-GcRF-XD8JYQi+OggTyMQ1C8RomdeQk3EsJwrjlOA3jiPUsiEIvbTRO0-T0PdXIjKY2SWIDdjtV1EyeNAtTSMEijbMsez7AM5zGJkvcbAUw8fJU8zIMswKtOQsSbAc0dDMikAd2iyAhBwkIlBEAAiSAsCUUqD3i5ggA)

if you check the last few rows, you can see the calls of the "hook".. if you pass an interface which only contains nullable properties, then also the initial-state itself gets optional. On the other side, if you pass an interface that has a mandatory property, also the initial-state gets mandatory...
and with this typing, the `getDefaultState` returns always a state that is not null

Reply:
@neotob Interesting, and I think it's a good solution :D 

@neotrow What do you think?Do you agree as well?

Reply:
@manni497 

i also think the solution of @neotob should work well :-) 

Thanks both of you 🙏

Reply:
@neotob @neotrow implemented :D

Reply:
change looks good to me 💪